File: app/src/classes/Student.js
/**
 * Student Class
 * Demonstrates: Inheritance, Encapsulation, and Polymorphism
 * Extends Person and implements GradeCalculator interface
 */

import { Person } from "./Person";
import { calculateGradeFromMarks, getDefaultGradeDescription } from "./GradeCalculator";

export class Student extends Person {
  // Private attributes - Encapsulation
  #course;
  #marks;
  #grade;

  constructor(id, name, course, marks) {
    // Call parent constructor - Inheritance
    super(id, name);
    this.#course = course;
    this.setMarks(marks); // Use setter for validation
  }

  // Getters - Encapsulation
  getCourse() {
    return this.#course;
  }

  getMarks() {
    return this.#marks;
  }

  getGrade() {
    return this.#grade;
  }

  // Setters - Encapsulation with validation
  setCourse(course) {
    this.#course = course;
  }

  setMarks(marks) {
    // Validation - Encapsulation ensures data integrity
    if (marks >= 0 && marks <= 100) {
      this.#marks = marks;
      this.#grade = this.calculateGrade(marks); // Recalculate grade
    } else {
      throw new Error(
        "Invalid marks! Marks must be between 0 and 100."
      );
    }
  }

  // Polymorphism: Override calculateGrade method from GradeCalculator
  calculateGrade(marks) {
    return calculateGradeFromMarks(marks);
  }

  // Polymorphism: Implement interface method
  getGradeDescription(grade) {
    return getDefaultGradeDescription(grade);
  }

  // Polymorphism: Override display method from Person abstract class
  display() {
    return {
      id: this.getId(),
      name: this.getName(),
      course: this.#course.toString(),
      marks: this.#marks,
      grade: this.#grade,
      gradeStatus: this.getGradeDescription(this.#grade),
      basicInfo: this.printBasicInfo()
    };
  }

  // Additional method to get full student profile
  displayFullProfile() {
    return {
      studentId: this.getId(),
      studentName: this.getName(),
      courseCode: this.#course.getCourseCode(),
      courseName: this.#course.getCourseName(),
      marksObtained: this.#marks,
      gradeAchieved: this.#grade,
      performance: this.getGradeDescription(this.#grade)
    };
  }

  // Get all student data as object
  toJSON() {
    return {
      id: this.getId(),
      name: this.getName(),
      course: this.#course.display(),
      marks: this.#marks,
      grade: this.#grade
    };
  }

  // Clone student object
  clone() {
    return new Student(
      this.getId(),
      this.getName(),
      this.#course,
      this.#marks
    );
  }
}

File: java-app/src/Student.java
/**
 * Student Class
 * Demonstrates: Inheritance, Encapsulation, and Polymorphism
 * Extends Person and implements GradeCalculator interface
 */
public class Student extends Person implements GradeCalculator {
    // Encapsulation: Private attributes
    private Course course;
    private double marks;
    private String grade;

    // Constructor
    public Student(String id, String name, Course course, double marks) {
        super(id, name); // Call parent constructor - Inheritance
        this.course = course;
        this.marks = marks;
        this.grade = calculateGrade(marks); // Calculate grade upon creation
    }

    // Getters - Encapsulation
    public Course getCourse() {
        return course;
    }

    public double getMarks() {
        return marks;
    }

    public String getGrade() {
        return grade;
    }

    // Setters - Encapsulation
    public void setCourse(Course course) {
        this.course = course;
    }

    public void setMarks(double marks) {
        if (marks >= 0 && marks <= 100) {
            this.marks = marks;
            this.grade = calculateGrade(marks); // Recalculate grade
        } else {
            System.out.println("Invalid marks! Marks must be between 0 and 100.");
        }
    }

    // Polymorphism: Override calculateGrade method from GradeCalculator interface
    @Override
    public String calculateGrade(double marks) {
        if (marks >= 90) {
            return "A";
        } else if (marks >= 80) {
            return "B";
        } else if (marks >= 70) {
            return "C";
        } else if (marks >= 60) {
            return "D";
        } else {
            return "F";
        }
    }

    // Polymorphism: Override display method from Person abstract class
    @Override
    public void display() {
        System.out.println("\n========== STUDENT DETAILS ==========");
        printBasicInfo(); // Call parent method
        System.out.println("Course: " + course.getCourseName() + " (" + course.getCourseCode() + ")");
        System.out.println("Marks: " + marks);
        System.out.println("Grade: " + grade);
        System.out.println("Grade Status: " + getGradeDescription(grade)); // Interface method
        System.out.println("====================================");
    }

    // Additional method to display full student profile
    public void displayFullProfile() {
        System.out.println("\n========== FULL STUDENT PROFILE ==========");
        System.out.println("Student ID: " + getId());
        System.out.println("Student Name: " + getName());
        course.display();
        System.out.println("Marks Obtained: " + marks + "/100");
        System.out.println("Grade Achieved: " + grade);
        System.out.println("Performance: " + getGradeDescription(grade));
        System.out.println("==========================================");
    }
}
